sdk/python/lib/pulumi/_types.py

```diff
@@ -0,0 +1,433 @@
+# limitations under the License.
+
+import builtins
+import functools
+import sys
+import typing
+from typing import Any, Callable, Dict, Optional, Type, TypeVar, Union, cast, get_type_hints
+
+from . import _utils
+
+T = TypeVar('T')
+
+
+_PULUMI_GETTER = "_pulumi_getter"
+_PULUMI_NAME = "_pulumi_name"
+_PULUMI_INPUT_TYPE = "_pulumi_input_type"
+_PULUMI_OUTPUT_TYPE = "_pulumi_output_type"
+_TRANSLATE_PROPERTY = "_translate_property"
+_VALUES = "_values"
+
+
+def is_input_type(cls: type) -> bool:
+    return hasattr(cls, _PULUMI_INPUT_TYPE)
+
+def is_output_type(cls: type) -> bool:
+    return hasattr(cls, _PULUMI_OUTPUT_TYPE)
+
+
+class _MISSING_TYPE:
+    pass
+MISSING = _MISSING_TYPE()
+"""
+MISSING is a singleton sentinel object to detect if a parameter is supplied or not.
+"""
+
+class _Property:
+    """
+    Represents a Pulumi property. It is not meant to be created outside this module,
+    rather, the property() function should be used.
+    """
+    def __init__(self, name: str, default: Any = MISSING) -> None:
+        if not name:
+            raise TypeError("Missing name argument")
+        if not isinstance(name, str):
+            raise TypeError("Expected name to be a string")
+        self.name = name
+        self.default = default
+        self.type: Any = None
+
+
+# This function's return type is deliberately annotated as Any so that type checkers do not
+# complain about assignments that we want to allow like `my_value: str = property("myValue")`.
+# pylint: disable=redefined-builtin
+def property(name: str, default: Any = MISSING) -> Any:
+    """
+    Return an object to identify Pulumi properties.
+
+    name is the Pulumi property name.
+    """
+    return _Property(name, default)
+
+
+def _properties_from_annotations(cls: type) -> Dict[str, _Property]:
+    """
+    Returns a dictionary of properties from annotations defined on the class.
+    """
+
+    # Get annotations that are defined on this class (not base classes).
+    cls_annotations = cls.__dict__.get('__annotations__', {})
+
+    def get_property(cls: type, a_name: str, a_type: Any) -> _Property:
+        default = getattr(cls, a_name, MISSING)
+        p = default if isinstance(default, _Property) else _Property(name=a_name, default=default)
+        p.type = a_type
+        return p
+
+    return {
+        name: get_property(cls, name, type)
+        for name, type in cls_annotations.items()
+    }
+
+
+def _process_class(cls: type, signifier_attr: str) -> Dict[str, Any]:
+    # Get properties.
+    props = _properties_from_annotations(cls)
+
+    # Clean-up class attributes.
+    for name in props:
+        # If the class attribute (which is the default value for this prop)
+        # exists and is of type 'Property', delete the class attribute so
+        # it is not set at all in the post-processed class.
+        if isinstance(getattr(cls, name, None), _Property):
+            delattr(cls, name)
```

Comment:
Can you explain this a bit more? I'll admit I'm a bit lost here :)

Reply:
I can embellish in a comment, but in the meantime, here's an explanation:

### Quick summary of dataclass feature

The design and implementation is inspired and based on the built-in dataclass feature, where you can define classes like:

```python
@dataclass
class C:
    x: int
    y: int = field(repr=False)
    z: int = field(repr=False, default=10)
    t: int = 20
```

Relevant quotes from https://docs.python.org/3/library/dataclasses.html#dataclasses.dataclass:

> The dataclass() decorator examines the class to find `field`s. A `field` is defined as class variable that has a [type annotation](https://docs.python.org/3/glossary.html#term-variable-annotation).
>
> [...]
>
> fields may optionally specify a default value, using normal Python syntax:
>
> ```python
> @dataclass
> class C:
>    a: int       # 'a' has no default value
>    b: int = 0   # assign a default value for 'b'
>```

And https://docs.python.org/3/library/dataclasses.html#dataclasses.field:

> For common and simple use cases, no other functionality is required. There are, however, some dataclass features that require additional per-field information. To satisfy this need for additional information, you can replace the default field value with a call to the provided [`field()`](https://docs.python.org/3/library/dataclasses.html#dataclasses.field) function. For example:
>
> ```python
> @dataclass
> class C:
>     mylist: List[int] = field(default_factory=list)
> 
> c = C()
> c.mylist += [1, 2, 3]
> ```
>
> [...]
>
> If the default value of a field is specified by a call to `field()`, then the class attribute for this field will be replaced by the specified `default` value. If no `default` is provided, then the class attribute will be deleted. The intent is that after the `dataclass()` decorator runs, the class attributes will all contain the default values for the fields, just as if the default value itself were specified. For example, after:
>
> ```python
> @dataclass
> class C:
>     x: int
>     y: int = field(repr=False)
>     z: int = field(repr=False, default=10)
>     t: int = 20
> ```
>
> The class attribute `C.z` will be `10`, the class attribute `C.t` will be `20`, and the class attributes `C.x` and `C.y` will not be set.

## Our design

We're allowing output types to be specified in a similar way:

```
@pulumi.output_type
class Foo:
    my_arg: str = pulumi.property("myArg")
    some_other_value: bool = pulumi.property("someOtherValue")
```

Dataclasses uses "fields" as its terminology, but for us, I'm referring to these as "properties". Our "properties" can similarly be defined as class variables with type annotations. Our `pulumi.property()` function is analogous to the `dataclasses.field()` function, and is used to provide additional information associated with the "property": in this case, the Pulumi camelCase name (pun, omg, 😆).
	
We first determine the class's "properties" in `_properties_from_annotations()` by looking at the class's `__annotations__` (which is the dict `{"my_arg": str, "some_other_value": bool}` for the example above) and then look up the class attribute value for each key in that dict. In this example, the value of the `my_arg` class attribute is an instance of `_Property(name="myArg")`, and the value of `some_other_value` is `_Property(name="someOtherValue")`. We do some more massaging, store the type annotation in the `_Property` instance, and then `_properties_from_annotations()` returns a dictionary of the class attribute names to `_Property` instances.

Then the decorator "cleans up" the class, by removing any of the class attributes that were set to `_Property` via a call to `pulumi.property()`.

Dataclasses does the same thing, cleaning up the class attributes that were set to values returned from calls to `field()`.